internal/crypto/crypto.go

```diff
@@ -157,3 +162,257 @@ func NewEncryptedString(id string, data []byte, keyID string, keyBase64URL strin
+            log.Printf("[ERROR] Invalid chain ID: %s", msg.ChainID)
+            return siws.ErrInvalidChainID
+        }
+    }
+
+    // 6) Nonce validation
+    if msg.Nonce != "" {
+        log.Printf("[DEBUG] Checking nonce length: %d", len(msg.Nonce))
+        if len(msg.Nonce) < 8 {
+            log.Printf("[ERROR] Nonce too short: %d chars", len(msg.Nonce))
+            return siws.ErrNonceTooShort
+        }
+    }
+
+    // 7) URI validation
+    if msg.URI != "" {
+        log.Printf("[DEBUG] Validating URI: %s", msg.URI)
+        if _, err := url.Parse(msg.URI); err != nil {
+            log.Printf("[ERROR] Invalid URI: %s - %v", msg.URI, err)
+            return siws.ErrInvalidURI
+        }
+    }
+
+    // Resources validation
+    for _, resource := range msg.Resources {
+        log.Printf("[DEBUG] Validating resource URI: %s", resource)
+        if _, err := url.Parse(resource); err != nil {
+            log.Printf("[ERROR] Invalid resource URI: %s - %v", resource, err)
+            return siws.ErrInvalidResourceURI
+        }
+    }
+
+    // 8) Signature verification
+    log.Printf("[DEBUG] Verifying ed25519 signature")
+	log.Printf("[DEBUG] Verification inputs - Message bytes: %v", []byte(rawMessage))
+	log.Printf("[DEBUG] Verification inputs - Signature bytes: %v", signature)
+    if !ed25519.Verify(pubKey, []byte(rawMessage), signature) {
+        log.Printf("[ERROR] Signature verification failed")
+        return siws.ErrSignatureVerification
+    }
+
+    // 9) Time validations
+    now := time.Now().UTC()
+    log.Printf("[DEBUG] Time validation - Current time: %s", now)
+
+    if !msg.IssuedAt.IsZero() {
+        log.Printf("[DEBUG] Checking issuedAt: %s", msg.IssuedAt)
+        if now.Before(msg.IssuedAt) {
+            log.Printf("[ERROR] Message from future - IssuedAt: %s", msg.IssuedAt)
+            return siws.ErrFutureMessage
+        }
+
+        if params.CheckTime && params.TimeDuration > 0 {
+            expiry := msg.IssuedAt.Add(params.TimeDuration)
+            log.Printf("[DEBUG] Checking message expiry - Expiry: %s", expiry)
+            if now.After(expiry) {
+                log.Printf("[ERROR] Message expired - Expiry: %s", expiry)
+                return siws.ErrMessageExpired
+            }
+        }
+    }
+
+    if !msg.NotBefore.IsZero() {
+        log.Printf("[DEBUG] Checking notBefore: %s", msg.NotBefore)
+        if now.Before(msg.NotBefore) {
+            log.Printf("[ERROR] Message not yet valid - NotBefore: %s", msg.NotBefore)
+            return siws.ErrNotYetValid
+        }
+    }
+
+    if !msg.ExpirationTime.IsZero() {
+        log.Printf("[DEBUG] Checking expirationTime: %s", msg.ExpirationTime)
+        if now.After(msg.ExpirationTime) {
+            log.Printf("[ERROR] Message expired - ExpirationTime: %s", msg.ExpirationTime)
+            return siws.ErrMessageExpired
+        }
+    }
+
+    log.Printf("[INFO] SIWS verification successful")
+    return nil
```

Comment:
Instead of checking nonce state, ensure that the timespan between `Issued At` and `Expires At` is within 5 minutes. This makes it very hard to reuse a signed message that was accidentally logged or stored in a compromised system. 

As end-user device compromise is not something we can guard against, someone having access to the signed message within 5 minutes of wallet use means bigger problems for the wallet holder.

Developers should use MFA instead to guard against wallet or end-user device compromise.